src/historywork/VerifyTxResultsWork.cpp

```diff
@@ -0,0 +1,157 @@
+            CLOG(TRACE, "History")
+                << "Transaction results verification for checkpoint "
+                << checkpoint
+                << (verified ? " successful"
+                             : (" failed: " +
+                                std::string(POSSIBLY_CORRUPTED_HISTORY)));
+            if (!verified)
+            {
+                ec = std::make_error_code(std::errc::io_error);
+            }
+
+            // Wake up call is posted back on the main thread
+            self->mApp.postOnMainThread(
+                [weak, ec]() {
+                    auto self = weak.lock();
+                    if (self)
+                    {
+                        self->mEc = ec;
+                        self->mDone = true;
+                        self->wakeUp();
+                    }
+                },
+                "VerifyTxResults: finish");
+        }
+    };
+
+    mApp.postOnBackgroundThread(verify, "VerifyTxResults: start in background");
+    return State::WORK_WAITING;
+}
+
+bool
+VerifyTxResultsWork::verifyTxResultsOfCheckpoint()
+{
+    try
+    {
+        FileTransferInfo hi(mDownloadDir, HISTORY_FILE_TYPE_LEDGER,
+                            mCheckpoint);
+        FileTransferInfo ri(mDownloadDir, HISTORY_FILE_TYPE_RESULTS,
+                            mCheckpoint);
+        mHdrIn.open(hi.localPath_nogz());
+        mResIn.open(ri.localPath_nogz());
+    }
+    catch (FileSystemException&)
+    {
+        CLOG(ERROR, "History") << "Transaction results failed verification: "
+                                  "are .xdr files downloaded?";
+        CLOG(ERROR, "History") << POSSIBLY_CORRUPTED_LOCAL_FS;
+        return false;
+    }
+
+    LedgerHeaderHistoryEntry curr;
+    while (mHdrIn && mHdrIn.readOne(curr))
+    {
+        auto ledgerSeq = curr.header.ledgerSeq;
+        auto txResultEntry = getCurrentTxResultSet(ledgerSeq);
+        auto resultSetHash =
+            sha256(xdr::xdr_to_opaque(txResultEntry.txResultSet));
+        auto genesis = ledgerSeq == LedgerManager::GENESIS_LEDGER_SEQ &&
+                       txResultEntry.txResultSet.results.empty();
+
+        if (!genesis && resultSetHash != curr.header.txSetResultHash)
+        {
+            CLOG(ERROR, "History") << fmt::format(
+                "Hash of result set does not agree with result "
+                "hash in ledger header: "
+                "txset result hash for {:d} is {:s}, but expected {:s}",
+                ledgerSeq, hexAbbrev(resultSetHash),
+                hexAbbrev(curr.header.txSetResultHash));
+            return false;
+        }
+    }
+    return true;
+};
+
+TransactionHistoryResultEntry
+VerifyTxResultsWork::getCurrentTxResultSet(uint32_t ledger)
+{
+    TransactionHistoryResultEntry trs;
+    trs.ledgerSeq = ledger;
+
+    do
+    {
+        if (mTxResultEntry.ledgerSeq < ledger)
+        {
+            CLOG(DEBUG, "History") << "Skipping tx results for ledger "
```

Comment:
I think this would be a bogus file (with extra results) ; for example: a file for ledgers [a..a+63] should not contain results for ledgers smaller than a, greater than a+64, and should not contain duplicates